File: pocketide/tests/capacity.py
#!/usr/bin/env python3
"""The build-memory ladder, in the two places it has to be identical.

pocketide-tools.sh WRITES the numbers into Gradle's own settings on the phone. Capacity.java
TELLS THE OWNER what was written. They are two copies of one table, in two languages, and the
failure mode when they drift is silent and confidence-destroying: a screen that says "2048 MB
build heap" beside a build that actually got 1024 and died.

There is no way to share one table between a bash script and a Java class without inventing a
build step to generate one from the other, which for five rows is worse than this test.

The check is the table, not the wording: every threshold, every heap and every worker count is
parsed out of both files and compared row by row.
"""
import re
import sys

app = sys.argv[1]
script = open(app + "/app/assets/pocketide-tools.sh").read()

# Comments stripped, and that matters more than it looks: Capacity.java's own header explains at
# length why getMemoryClass() is the wrong number to quote, and a naive search for the word finds
# the explanation and fails the file for containing its own reasoning.
java = open(app + "/app/src/com/pocketide/Capacity.java").read()
java = re.sub(r'/\*.*?\*/', '', java, flags=re.S)
java = re.sub(r'//.*$', '', java, flags=re.M)

problems = []

# --- the shell ladder ------------------------------------------------------------------------
# elif [ "$total_kb" -ge 7500000  ]; then heap=2048; workers=3; kotlin_heap=1024
shell_rows = []
for match in re.finditer(
        r'\[\s*"\$total_kb"\s*-ge\s*(\d+)\s*\]\s*;\s*then\s+heap=(\d+);\s*workers=(\d+)',
        script):
    shell_rows.append((int(match.group(1)), int(match.group(2)), int(match.group(3))))

shell_default = re.search(r'else\s+heap=(\d+);\s*workers=(\d+)', script)
if not shell_rows or not shell_default:
    problems.append("the build-memory ladder cannot be read out of pocketide-tools.sh")

# --- the Java ladder -------------------------------------------------------------------------
# if (totalRam >= 7_500_000L * 1024) return 2048;
heap_body = re.search(r'static int buildHeapMb\(long totalRam\)\s*\{(.*?)\n    \}', java, re.S)
worker_body = re.search(r'static int workers\(long totalRam, int cores\)\s*\{(.*?)\n    \}',
                        java, re.S)

java_heap = []
java_heap_default = None
if not heap_body:
    problems.append("Capacity.buildHeapMb cannot be read")
else:
    for match in re.finditer(r'totalRam >= ([\d_]+)L \* 1024\) return (\d+);', heap_body.group(1)):
        java_heap.append((int(match.group(1).replace("_", "")), int(match.group(2))))
    tail = re.search(r'return (\d+);\s*$', heap_body.group(1).strip())
    if tail:
        java_heap_default = int(tail.group(1))

java_workers = []
java_workers_default = None
if not worker_body:
    problems.append("Capacity.workers cannot be read")
else:
    for match in re.finditer(r'totalRam >= ([\d_]+)L \* 1024\) byMemory = (\d+);',
                             worker_body.group(1)):
        java_workers.append((int(match.group(1).replace("_", "")), int(match.group(2))))
    tail = re.search(r'else byMemory = (\d+);', worker_body.group(1))
    if tail:
        java_workers_default = int(tail.group(1))

# --- compare ---------------------------------------------------------------------------------
if shell_rows and java_heap:
    if [(kb, heap) for kb, heap, _ in shell_rows] != java_heap:
        problems.append(
            "the build heap ladder differs. pocketide-tools.sh writes %s; Capacity.java tells "
            "the owner %s. A screen that promises more memory than the build gets is worse "
            "than saying nothing."
            % ([(kb, heap) for kb, heap, _ in shell_rows], java_heap))

if shell_default and java_heap_default is not None:
    if int(shell_default.group(1)) != java_heap_default:
        problems.append("the smallest-phone build heap is %s in the script and %s in Capacity"
                        % (shell_default.group(1), java_heap_default))

# The worker ladders are allowed to differ in SHAPE but not in outcome: Java also caps by core
# count, which the script does not need to because Gradle reads the same machine. So every
# threshold the script names must produce the same worker count in Java for a phone with
# enough cores -- which is what the comparison below does.
if shell_rows and java_workers:
    java_at = dict(java_workers)
    for kb, _, workers in shell_rows:
        if kb in java_at and java_at[kb] != workers:
            problems.append("at %d kB of RAM the script uses %d parallel worker(s) and "
                            "Capacity says %d" % (kb, workers, java_at[kb]))
        elif kb not in java_at and workers != max(w for _, w in java_workers):
            # A threshold the script has and Java does not is only safe when the script's
            # answer there is one Java also reaches.
            if workers not in [w for _, w in java_workers] + [java_workers_default]:
                problems.append("the script uses %d worker(s) at %d kB, a count Capacity never "
                                "reports for any phone" % (workers, kb))

# --- the honest claims the screen makes ------------------------------------------------------
#
# Two are load-bearing and were both nearly got wrong here. getMemoryClass() bounds the app's
# own Dalvik heap and has nothing to do with a PRoot child process; quoting it would have told
# every owner their computer had 256 MB. And the APK's own size is not what limits a build --
# saying otherwise sends someone off to buy a phone they did not need.
if "getMemoryClass" in java:
    problems.append("Capacity quotes ActivityManager.getMemoryClass, which bounds the app's own "
                    "Dalvik heap and not the PRoot processes the workspace actually runs in")
if "availableProcessors" not in java:
    problems.append("Capacity never reads the core count it reports")
if "NDK" not in java:
    problems.append("Capacity does not mention that C and C++ cannot be built, which is the "
                    "limit an owner is most likely to hit after spending an afternoon")
if "Android 17" not in java:
    problems.append("Capacity does not mention the Android 17 memory limiter, which is the one "
                    "hard ceiling on how large a build can get")

for problem in problems:
    print("  " + problem, file=sys.stderr)
sys.exit(1 if problems else 0)

File: pocketide/app/src/com/pocketide/Capacity.java
package com.pocketide;

import android.app.ActivityManager;
import android.content.Context;
import android.os.Build;
import android.os.StatFs;

import java.util.Locale;

/**
 * What this computer actually is, in the numbers of the phone it is running on.
 *
 * "Ubuntu on your phone" tells an owner nothing about whether their project will build. The
 * questions they really have are how much memory it gets, how many cores, how much room, and
 * how big a thing they can build before it stops working -- and every one of those has a real
 * answer on THIS phone rather than a general one.
 *
 * Two of the answers are counter-intuitive enough to be worth stating plainly, because getting
 * them wrong is what sends someone off to buy a phone they did not need:
 *
 *   THE APK'S SIZE IS NOT THE LIMIT. A 200 MB APK is no harder to produce than a 2 MB one --
 *   it is bytes through a zip. What costs memory is the compiler: how many modules, how many
 *   source files, how large the dependency graph, and whether R8 has to rewrite the whole
 *   program at the end. A small app with two hundred dependencies is a heavier build than a
 *   large app with ten.
 *
 *   THERE IS NO FIXED ALLOCATION. Nothing here reserves 2 GB or 4 GB. Linux gets what Android
 *   is not using, moment to moment, and what takes it away is another app being opened. The one
 *   hard ceiling is the Android 17 memory limiter, which is per-app, derived from the device's
 *   total RAM, and cannot be opted out of -- see Exits.java, which is what reports it when it
 *   fires.
 *
 * The Java heap limit an Android developer would reach for -- ActivityManager.getMemoryClass()
 * -- is deliberately NOT used here, and it is the obvious mistake to make. It bounds the app's
 * own Dalvik heap. Everything in the workspace is a separate native process started through
 * PRoot, with its own address space, and none of them are inside that heap. Quoting it would
 * have told every owner their computer had 256 MB.
 */
final class Capacity {

    /** Everything the screen needs, measured rather than assumed. */
    static final class Reading {
        final int cores;
        final long totalRam;
        final long availableRam;
        final long freeStorage;
        final long workspaceBytes;      // -1 until it has been measured
        final int buildHeapMb;
        final int workers;

        Reading(int cores, long totalRam, long availableRam, long freeStorage,
                long workspaceBytes, int buildHeapMb, int workers) {
            this.cores = cores;
            this.totalRam = totalRam;
            this.availableRam = availableRam;
            this.freeStorage = freeStorage;
            this.workspaceBytes = workspaceBytes;
            this.buildHeapMb = buildHeapMb;
            this.workers = workers;
        }
    }

    private Capacity() {}

    /** Cheap: no PRoot, no disk walk. Safe on the drawing thread. */
    static Reading read(Context context) {
        ActivityManager am = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        ActivityManager.MemoryInfo memory = new ActivityManager.MemoryInfo();
        if (am != null) am.getMemoryInfo(memory);

        StatFs storage = new StatFs(context.getFilesDir().getAbsolutePath());
        int cores = Math.max(1, Runtime.getRuntime().availableProcessors());
        return new Reading(cores, memory.totalMem, memory.availMem, storage.getAvailableBytes(),
                -1, buildHeapMb(memory.totalMem), workers(memory.totalMem, cores));
    }

    /**
     * The build heap this phone gets, matching pocketide-tools.sh exactly.
     *
     * The same ladder lives in both places because both have to be true at once: the script
     * writes the number into Gradle's settings, and this screen tells the owner what was
     * written. A gate compares them, so they cannot drift apart silently.
     */
    static int buildHeapMb(long totalRam) {
        if (totalRam >= 11_500_000L * 1024) return 3072;
        if (totalRam >= 7_500_000L * 1024) return 2048;
        if (totalRam >= 5_500_000L * 1024) return 1536;
        if (totalRam >= 3_500_000L * 1024) return 1024;
        return 768;
    }

    static int workers(long totalRam, int cores) {
        int byMemory;
        if (totalRam >= 11_500_000L * 1024) byMemory = 4;
        else if (totalRam >= 7_500_000L * 1024) byMemory = 3;
        else if (totalRam >= 3_500_000L * 1024) byMemory = 2;
        else byMemory = 1;
        return Math.max(1, Math.min(byMemory, cores));
    }

    /** One line for the row: cores, memory, room. */
    static String oneLine(Reading reading) {
        return reading.cores + (reading.cores == 1 ? " core · " : " cores · ")
                + DeviceProbe.formatBytes(reading.totalRam) + " RAM · "
                + DeviceProbe.formatBytes(reading.freeStorage) + " free";
    }

    /**
     * The full account, written for someone deciding whether to start a real project here.
     *
     * Every number in it is either measured from this phone or published by the people who
     * decide it. Nothing is a benchmark, because a benchmark run on one phone and quoted for
     * every phone is the kind of claim that makes the rest of a screen untrustworthy.
     */
    static String describe(Context context, Reading reading) {
        StringBuilder words = new StringBuilder();

        words.append("THIS PHONE, RIGHT NOW\n\n")
                .append("Processor · ").append(reading.cores)
                .append(reading.cores == 1 ? " core" : " cores")
                .append(", ").append(Build.SUPPORTED_ABIS.length == 0
                        ? "arm64" : Build.SUPPORTED_ABIS[0])
                .append(". Linux sees every one of them. PRoot translates system calls; it does "
                        + "not emulate a processor, so code here runs at the phone's own speed "
                        + "rather than a fraction of it.\n\n")

                .append("Memory · ").append(DeviceProbe.formatBytes(reading.totalRam))
                .append(" in the phone, ").append(DeviceProbe.formatBytes(reading.availableRam))
                .append(" free at this moment. None of it is reserved for Linux: it takes what "
                        + "Android is not using, and gives it back when another app wants it. "
                        + "Closing a few apps before a long build is the single most effective "
                        + "thing you can do.\n\n")

                .append("Storage · ").append(DeviceProbe.formatBytes(reading.freeStorage))
                .append(" free. Linux lives inside this app's own storage on the phone's "
                        + "internal drive — not on an SD card, which Android does not allow an "
                        + "app to use this way. Uninstalling PocketIDE deletes all of it.\n\n")

                .append("Build memory · ").append(reading.buildHeapMb).append(" MB, with ")
                .append(reading.workers).append(reading.workers == 1
                        ? " module compiled at a time" : " modules compiled at a time")
                .append(". Both are worked out from the memory above and written into Gradle's "
                        + "own settings when the Android build tools are installed. They can be "
                        + "changed in ~/.gradle/gradle.properties like on any other machine.\n\n")

                .append("HOW BIG A PROJECT CAN BE BUILT\n\n")

                .append("The APK's size is not the limit. A 200 MB APK is no harder to produce "
                        + "than a 2 MB one — it is bytes through a zip file. What costs memory "
                        + "is the compiler, and what decides that is the number of modules, the "
                        + "number of source files, the size of the dependency graph, and "
                        + "whether R8 has to rewrite the whole program at the end. A small app "
                        + "with two hundred dependencies is a heavier build than a large app "
                        + "with ten.\n\n")

                .append("So the honest answer is a shape rather than a number:\n\n")

                .append("· A single-module app — one screen or fifty — builds comfortably on "
                        + "any phone this app will install on.\n")
                .append("· A multi-module app builds fine; each extra module in parallel wants "
                        + "its own memory, which is why the worker count above is capped.\n")
                .append("· A release build with R8 minification is the heaviest step in any "
                        + "Android build, and the one most likely to need more heap than a "
                        + "phone with 4 GB can give. Building debug first, and release when "
                        + "the phone is otherwise idle, is what makes it finish.\n")
                .append("· Anything with C or C++ in it cannot be built at all, on any phone. "
                        + "Google publishes no arm64 Android NDK. That is their decision, not a "
                        + "limit of this hardware.\n\n")

                .append("Websites, servers, command-line programs and anything else that "
                        + "compiles for arm64 Linux have no such ceiling. Node, Deno, Python, "
                        + "Go, Rust, Ruby, PHP, C and C++ all build here natively, at full "
                        + "speed, for this machine.\n\n")

                .append("WHAT ACTUALLY STOPS A BUILD\n\n")

                .append("Memory, and specifically Android taking it back. From Android 17 "
                        + "(June 2026) there is a per-app memory ceiling derived from the "
                        + "device's total RAM, applied whatever an app targets, with no way to "
                        + "opt out. A workspace running the editor, an extension host and a "
                        + "compiler is exactly the shape it is aimed at. When it fires, "
                        + "PocketIDE reads Android's own record and says so on the Home screen "
                        + "rather than leaving you to guess.\n\n")

                .append("Heat, second. A phone has no fan. A long build raises the temperature "
                        + "until Android slows the processor down to protect it — the build "
                        + "still finishes, it just takes longer. The Home screen shows the "
                        + "phone's thermal state for this reason.\n\n")

                .append("Battery, third. A twenty-minute build with the screen off is exactly "
                        + "what a phone's battery manager exists to stop. The permission rows "
                        + "in Settings are the switches that let it continue.\n\n")

                .append("WHAT MAKES IT FASTER\n\n")

                .append("· Plug the phone in. Charging relaxes both the battery manager and, on "
                        + "most phones, the processor's own governor.\n")
                .append("· Close other apps before a long build, for the memory.\n")
                .append("· Leave Gradle's build cache on — it is on by default here — so the "
                        + "second build of a project does a fraction of the work of the first.\n")
                .append("· Keep the phone out of a case and off a bed or sofa. Heat is the "
                        + "limit long before the processor is.\n");

        return words.toString();
    }

    /** How the phone compares to what a project's own documentation usually assumes. */
    static String shortVerdict(Reading reading) {
        long gb = Math.round(reading.totalRam / 1e9);
        if (gb >= 12) return "Large projects, including release builds with minification";
        if (gb >= 8) return "Most real projects, including multi-module ones";
        if (gb >= 6) return "Real projects; release builds want the phone otherwise idle";
        if (gb >= 4) return "Single-module apps and web projects comfortably";
        return String.format(Locale.ROOT, "Small projects; %d GB is tight for a compiler", gb);
    }
}
